### PR TITLE
Add Invalid Encoding Channel Warning

### LINKED ISSUE #3415: Invalid encoding channel should fail more gracefully
For example, if I type `_y` instead of `y`, everything crashes.

```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v2.json",
  "data": {"url": "data/movies.json"},
  "mark": "bar",
  "encoding": {
    "x": {
      "bin": true,
      "field": "IMDB_Rating",
      "type": "quantitative"
    },
    "_y": {
      "aggregate": "count",
      "type": "quantitative"
    }
  }
}
```

image

### PR DESCRIPTION
**Pull request by invokesus**
Please:
- [X] Make your pull request atomic, fixing one issue at a time unless there are many relevant issues that cannot be decoupled.
- [x] Provide a test case & update the documentation under `site/docs/`
- [X] Make lint and test pass. (Run `yarn lint` and `yarn test`.  If your change affects Vega outputs of some examples, please run `yarn build:example EXAMPLE_NAME` to re-compile a specific example or `yarn build:examples` to re-compile all examples.)
- [X] Make sure you have rebased onto the `master` branch.
- [X] Provide a concise title so that we can just copy it to our release note.
  - Use imperative mood and present tense.
  - Mention relevant issues. (e.g., `#1`)

Fixes #3415.

### CODE REVIEW

**Comment by domoritz** [comment]
Thank you for the pull request.

Please make this a warning and delete the encoding instead of throwing an error. Also, let's add a test.

**Comment by invokesus** [comment]
The warning and deletion of encoding is already implemented in [encoding.ts](https://github.com/vega/vega-lite/blob/ecc47de34e2603a87540bd2e70059225236bd4e6/src/encoding.ts#L146-L153).
I was thinking of adding a case for unrecognized Channel [here](https://github.com/vega/vega-lite/blob/ecc47de34e2603a87540bd2e70059225236bd4e6/src/channel.ts#L216). So that, `getSupportedMark` returns an empty dictionary, causing `supportMark` to return false and finally the warning being triggered in `normalizeEncoding`. 
Is this in the right direction?

**Comment by domoritz** [comment]
I'd trace what causes the error in #3415 and make sure that it's handled either by the code you found or something new you write.

**Comment by invokesus** [comment]
Updated the PR. 🙂

File: src/channel.ts [hunk 1]
```diff
@@ -214,6 +214,8 @@ export function getSupportedMark(channel: Channel): SupportedMark {
       return {point: true, geoshape: true};
     case TEXT:
       return {text: true};
+    default:
```
Comment (kanitw) [inline src/channel.ts]: I wonder if this is the right place to correct this?  

If the channel you have isn't a valid `Channel` at all, then it doesn't satisfy the signature of this method. 

A better check would be using `isChannel` method to check if a key of the encoding object  is a channel.

File: test/compile/unit.test.ts [hunk 2]
```diff
@@ -19,6 +19,18 @@ describe('UnitModel', function() {
...
 
+    it('should drop invalid channel and throws warning', log.wrap((localLogger) => {
+        const INVALID_CHANNEL = '_y';
+        const model = parseUnitModel({
+          mark: 'bar',
+          encoding: {
+            _y: {type: 'quantitative'}
+          }
+        } as any); // To make parseUnitModel accept the model with invalid encoding channel
+        assert.equal(model.encoding.shape, undefined);
```
Comment (kanitw) [inline test/compile/unit.test.ts]: I'm not sure why this is related to `model.encoding.shape`.

Comment (invokesus) [reply test/compile/unit.test.ts]: Removed the assert.

File: src/channel.ts (hunk 1 above)
Comment (invokesus) [reply src/channel.ts]: Made the changes. Thanks for the help!

File: src/log.ts [hunk 3]
```diff
@@ -175,6 +175,10 @@ export namespace message {
     return `${channel} dropped as it is incompatible with "${markOrFacet}"${when ? ` when ${when}` : ''}.`;
   }
 
+  export function invalidEncodingChannel(encoding: string) {
```
Comment (kanitw) [inline src/log.ts]: `encoding` normally refer to encoding mapping, which is an object.

Rename to `channel` would be less misleading.

File: src/log.ts [hunk 4]
```diff
@@ -175,6 +175,10 @@ export namespace message {
...
+    return `${encoding} dropped as it is invalid.`;
```
Comment (kanitw) [inline src/log.ts]: `${channel}-encoding is dropped as ${channel} is not a valid encoding channel.`

File: test/compile/unit.test.ts [hunk 5]
```diff
@@ -19,6 +19,17 @@ describe('UnitModel', function() {
         assert.equal(localLogger.warns[0], log.message.incompatibleChannel(SHAPE, BAR));
       }));
 
+    it('should drop invalid channel and throws warning', log.wrap((localLogger) => {
+        const INVALID_CHANNEL = '_y';
```
Comment (kanitw) [inline test/compile/unit.test.ts]: remove this line and just use `"_y"` below?

File: src/log.ts (hunk 3 above)
Comment (invokesus) [reply src/log.ts]: Done.

**Comment by domoritz** [comment]
Thank you @invokesus!

**Pull request merged**